Draft: Web - Firebase Google Sign in and Sign out

criszz77 opened the pull request:
I'm not very experienced with Firebase, but I'm trying to figure out a nice way to integrate the rest of the sign in options for the web

I don't think we need any third party libraries to handle authorisations like Google SignIn. Firebase got us covered with everything we need in the browser. I also think it would be much easier than supporting react-native-google-signin for the web. (I tried it and I really don't appreciate the outcome)

I'm following this guide: https://firebase.google.com/docs/auth/web/google-signin and it seems to be working?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-firebase-authentication-example?pullRequest=6) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-firebase-authentication-example?pullRequest=6) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-firebase-authentication-example?pullRequest=6) it.</sub>

mikehardy commented:
Oh this is cool!

I agree with the approach of using firebase stuff in *this* context (where we know we have firebase available) vs trying to port react-native-google-signin and react-native-google-fbsdk to web, that sounds like no fun even if it would support a broader userbase (all users of those packages, vs just users of firebase social auth)

Some thoughts:
- it's all popup method here, but the user could be on a mobile device, and per docs: `The redirect method is preferred on mobile devices.` - could be a follow-on commit to integrate what appears to be the popular solution https://github.com/duskload/react-device-detect (which appears to be doing a User-Agent parse and thus should work "in general" as a way to choose)

- the shim code feels like it's in the wrong place. Is there a way to put this in the shim for google-signin ? My idea for the shims is that they are nothing except API implementations of existing packages and should serve directly as aliases in craco config via webpack aliasing, whereas this new shim is all new code and referenced directly. Seems like this signIn/signOut code could live here? https://github.com/invertase/react-native-firebase-authentication-example/blob/main/template/src/shims/google-signin-web.ts#L14 (and below)

criszz77 commented:
> * it's all popup method here, but the user could be on a mobile device, and per docs: `The redirect method is preferred on mobile devices.` - could be a follow-on commit to integrate what appears to be the popular solution https://github.com/duskload/react-device-detect (which appears to be doing a User-Agent parse and thus should work "in general" as a way to choose)

I'll have a look into that. The PR is still a draft. 

> * the shim code feels like it's in the wrong place. Is there a way to put this in the shim for google-signin ? My idea for the shims is that they are nothing except API implementations of existing packages and should serve directly as aliases in craco config via webpack aliasing, whereas this new shim is all new code and referenced directly. Seems like this signIn/signOut code could live here? https://github.com/invertase/react-native-firebase-authentication-example/blob/main/template/src/shims/google-signin-web.ts#L14 (and below)

I tried implementing GAPI in the shim for google-signin here: https://github.com/criszz77/react-native-firebase-authentication-example/blob/google-signin-web/template/src/shims/google-signin-web.ts

After I finished implementing some of the shim for the google-signin, I realised GAPI is deprecated: 
Source: https://developers.google.com/identity/sign-in/web/sign-in
![image](https://user-images.githubusercontent.com/55203625/144639137-b909b267-ab7a-44f3-8b78-eab1b88779a4.png)

I should try again using GSI, the new Google Identity Services: https://developers.google.com/identity/gsi/web

But then again, I realised this is a Firebase example, and the Firebase implementation was the easiest. However it felt a bit wrong to add this Firebase functionality in a shim for react-native-google-signin, because it conflicted with Firebase here: https://github.com/invertase/react-native-firebase-authentication-example/blob/f5559e38eb1f4ef64f7d7498ffeee4a882dfbec1/template/src/app/auth-providers/Google.tsx#L46

mikehardy commented:
I see what you're saying about the google sign in stuff - a mix of deprecation and slightly different flow leading to the differences.

Perhaps the shim code could be in firebase-auth-web, exposing the real firebase-js-sdk APIs just with a big warning that they are web only in the shim and need platform detection / conditional in your code to work. react-native-firebase could actually take a PR that adds them at the native level but throws and is documented that they are web only and only there as stubs to maintain API compatibility (in that they exist...) with firebase-js-sdk. 

Then a future extension there could be for react-native-firebase to have an optional peer dependency on react-native-google-signin to actually make the calls or something, but not doing that now

criszz77 commented:
I tried implementing the new `GSI` from Google for `react-native-google-signin`, and for some reason I find it harder than the deprecated `GAPI`. While I'm far away for having something stable, you can see the progress here: https://github.com/criszz77/react-native-firebase-authentication-example/blob/gsi-google-signin-web/template/src/shims/google-signin-web.ts

As you can see, using the Firebase Google Sign in is much simpler so I'll probably continue with having a better implementation for this one.

mikehardy commented:
> https://github.com/criszz77/react-native-firebase-authentication-example/blob/gsi-google-signin-web/template/src/shims/google-signin-web.ts

:scream:  oh my, that is a lot. I have a pretty strict policy of voicing opinions *but* following a rule I call "implementor's choice". In the end, if you are the implementor and you really tried a few things and most ways suck, but one works, as the implementor, you have the authority to say, "no really, this way is better even if XYZ".

So perhaps your original shim with platform guards is as good as it gets. I'd favor moving on then since it works, and spending your time creating whole new value on a different feature (here, or elsewhere :-) )

criszz77 commented:
All right, so I cleaned up a bit my changes and I also implemented the SignInWithRedirect for mobile using [`react-device-detect`](https://www.npmjs.com/package/react-device-detect)

It seems to me that adding web support for other providers like Facebook and Apple will be very easy using Firebase.

Please have a look and let me know what you think.

mikehardy commented on template/src/app/signed-in/Settings.tsx:
```diff
@@ -60,8 +60,13 @@ function EditProfile(): JSX.Element | null {
   async function signOut() {
     setSigningOut(true);
-    await GoogleSignin.signOut();
     await auth().signOut();
+
+    if (Platform.OS === 'web') {
+      await googleWebSignOut();
+    } else {
+      await GoogleSignin.signOut();
+    }
```
I think we need to sign out all providers before signing out auth(), otherwise the native SDKs do the auth signout, an auth changed event is fired, a listener is called async, and the user disappears, which causes a main App.tsx re-render which unmounts all these components in a race condition

criszz77 replied:
Fixed.

mikehardy commented on template/src/app/signed-out/SignIn.tsx:
```diff
@@ -51,7 +51,7 @@ function SignIn() {
 
         {Platform.OS !== 'web' && <Facebook />}
-        {Platform.OS !== 'web' && <Google />}
+        <Google />
```
Just want to say that everything else is important to make this happen, but this: this is the cool part :sunglasses:

mikehardy commented on template/src/app/types/@react-native-firebased.d.ts:
```diff
@@ -0,0 +1,6 @@
+import {UserCredential} from 'firebase/auth';
```
Hmm :thinking: the naming on this one `firebase**d**` ? Also, how does this interact with the other types from auth?

criszz77 replied:
Module Augmentation from typescript: https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation

Basically it merges those types to the existing ones.

criszz77 replied:
Fixed the typo and moved `types` folder to `src`

mikehardy replied:
cool - in fact module augmentation is how the whole firebase-js-sdk v8 (not v9!) type system worked
works for me :-)

mikehardy commented:
Agreed easy with firebase - sorry the attempt at doing it other ways went off the rails so badly :-)! I appreciate the attempt but hey, we have firebase here, no need to try to break through a wall of problems when there is a door right there right?

I left a couple other comments but other than a doc change to mention google signin for web is provided now

I would add a warning the API is only 2 calls and works great but is specific to this shim, needs a Platform guard, and may change. Just to set expectations on breaking changes so people aren't upset if we figure out a really clean way to do it that can hide behind the existing auth apis from the intersection of firebase-js-sdk and @react-native-firebase/auth (you never know)

mikehardy reviewed (approved):
honestly, I know you still have it marked for draft - and I was mostly "reviewing" in the hope of just trying to help / think out loud vs a formal review (whatever that is ;-) ) - but at this point it's in "looks good to me" status... let me know when you're happy with it and I can re-review if needed but if it's working in your testing, I could merge

criszz77 commented:
I'll have a final look a bit later and I will let you know. Thanks!

criszz77 commented:
For the web it seems to be ok, but I didn’t fully test on mobile because I continously had issues with the setup (I’m pretty bad with native code). If you have the chance, please test before merging. Thanks!

criszz77 commented:
Finally managed to set it up and test it properly on mobile as well. 👌

Will start working on Faceboook for web soon enough.

mikehardy commented:
Oh - sorry I pushed conflicts in to this, honestly I was just testing it locally on the main branch prior to reviewing your PRs and noticed a couple of major issues I needed to cleanup first (auth errors were not popping an alert on web, and they weren't translated either). Anyway, I will review these and merge them as soon as I have the main branch free of known bugs again

criszz77 commented:
Don't worry, let me know if there is anything I can help with.

criszz77 commented:
Hey, sorry I was a bit inactive, but holidays are coming, that means more time for open source. 😂 Are you still interested in those small changes? Should I resolve the conflicts?

mikehardy commented:
I still have one "urgent" work project I need to deliver before the holidays, so I haven't been pushing many changes into the generic / open layers myself! No need to apologize, of course, open source is open source, no one can make you feel bad

Well, there are people that don't contribute *and* they complain. I do my best to make those people feel bad but that's a personal choice and I'm "retired" so I don't care if I have one less "friend" ;-).

Anyway, very interested in getting these in actually, they are on my high priority list, unfortunately my "urgent" list still has items :weary:
